### PR TITLE
Verify Release URLs using publish attestations

### PR DESCRIPTION
**Pull request by facutuesca**
This PR adds a new field to `Release`, `trusted_publisher_url`, which stores the publisher URL used to create that release. 

More specifically, this field is only populated when the first file uploaded (the one that creates the release) is uploaded using Trusted Publishing **and** has a PEP-740 publish attestation.

This allows PyPI to verify a release's URLs (the ones returned by the `Release.urls` property), by comparing them with the `trusted_publisher_url`. This means we put those URLs in the `Verified details` section of the project page:
image

(compare with the mockup provided in this comment: https://github.com/pypi/warehouse/issues/8635#issuecomment-2008164863)

This is an alternative implementation to these two PRs:
https://github.com/pypi/warehouse/pull/15862
https://github.com/pypi/warehouse/pull/15891

cc @woodruffw

### CODE REVIEW

File: tests/unit/packaging/test_models.py [hunk 1]
```diff
@@ -555,6 +555,36 @@ def test_urls(self, db_session, home_page, download_url, project_urls, expected)
...
+                True,
+            ),
+            ("https://github.com/owner/", "https://github.com/owner/project", False),
+            (
+                "https://gitlab.com/owner/project",
+                "https://github.com/owner/project",
+                False,
+            ),
+        ],
+    )
```
Comment (woodruffw) [inline tests/unit/packaging/test_models.py:583]: Nitpick: might be good to have some ActiveState URL examples in here as well 🙂

File: warehouse/forklift/legacy.py [hunk 2]
```diff
@@ -1139,6 +1141,11 @@ def file_upload(request):
                         f"Attestation with unsupported predicate type: "
                         f"{predicate_type}",
                     )
+                elif is_new_release:
+                    # If this is a publish attestation and we are creating the release,
+                    # we store the signed URL of the repository, so that we can later
+                    # verify the URLs in the project.
+                    release.trusted_publisher_url = publisher.publisher_url()
```
Comment (woodruffw) [inline warehouse/forklift/legacy.py:1148]: NB: `publisher_url()` will be `None` for e.g. Google-based Trusted Publishers, so we should make sure that doesn't cause issues here or elsewhere.

File: warehouse/packaging/models.py [hunk 3]
```diff
@@ -653,9 +656,23 @@ def urls(self):
                 continue
 
             _urls[name] = url
-
         return _urls
 
+    def verify_url(self, url: str) -> bool:
+        if not self.trusted_publisher_url:
+            return False
+        return url.startswith(self.trusted_publisher_url)
```
Comment (woodruffw) [inline warehouse/packaging/models.py:664]: NB: This doesn't do case normalization or anything else, which is _probably_ fine but might be worth calling out in a comment.

**Comment by facutuesca** [comment]
@woodruffw While looking into URL normalization, I realized this implementation has a pretty bad security bug:
```python
publisher_url = "https://github.com/org/project"
release_url = "https://github.com/org/project22"
verify_url(release_url) == True  # because release_url.startswith(publisher_url) is True
```

Which led me to look into [`urllib`](https://docs.python.org/3/library/urllib.parse.html) to see if we can parse and compare URL elements. But `urllib` [doesn't do any validation on inputs](https://docs.python.org/3/library/urllib.parse.html#url-parsing-security), which means there's no guarantee a malicious input will not parse as a valid URL that matches the publisher URL.

Any ideas on how we can tackle URL comparison here?

**Comment by woodruffw** [comment]
> Which led me to look into [`urllib`](https://docs.python.org/3/library/urllib.parse.html) to see if we can parse and compare URL elements. But `urllib` [doesn't do any validation on inputs](https://docs.python.org/3/library/urllib.parse.html#url-parsing-security), which means there's no guarantee a malicious input will not parse as a valid URL that matches the publisher URL.

Yeah, I think we should avoid `urllib` entirely for this, and use a more standards-conformant URL parser that offers strong validation. One good candidate is [`rfc3986`](https://pypi.org/project/rfc3986/) -- technically that'll be stricter than WHATWG requires for URL validation, but this shouldn't be an issue in practice (and when it is, it'll be a good validation check).

TL;DR: I think we should use `rfc3986` + test the bejeezus out of it, including negative and backstop cases 🙂

File: tests/unit/packaging/test_models.py (hunk 1 above)
Comment (facutuesca) [reply tests/unit/packaging/test_models.py]: Done!

File: warehouse/forklift/legacy.py (hunk 2 above)
Comment (facutuesca) [reply warehouse/forklift/legacy.py:1148]: I checked, I believe it's fine since `Release.trusted_publisher_url` is already a `Optional[str]` (so we get type checking), and the only place where we use it is in the `Release.verify_url` method (which has an explicit `is None` path)

File: warehouse/packaging/models.py (hunk 3 above)
Comment (facutuesca) [reply warehouse/packaging/models.py]: With the new implementation, now we do normalization. I added a detailed docstring to document the behavior

**Comment by facutuesca** [comment]
> TL;DR: I think we should use `rfc3986` + test the bejeezus out of it, including negative and backstop cases 🙂

@woodruffw Done! The `Release.verify_url` now does more exhaustive checks, and I added several more tests for the error cases I could think of.

File: warehouse/packaging/models.py [hunk 4]
```diff
@@ -22,6 +22,7 @@
 from github_reserved_names import ALL as GITHUB_RESERVED_NAMES
 from pyramid.authorization import Allow, Authenticated
 from pyramid.threadlocal import get_current_request
+from rfc3986 import api
```
Comment (woodruffw) [inline warehouse/packaging/models.py:25]: Nitpick: `api` is a pretty generic import, so maybe we do `import rfc3986.api` instead and refer to things with that fully qualified path 🙂

File: warehouse/packaging/models.py [hunk 5]
```diff
@@ -653,9 +657,47 @@ def urls(self):
...
+        or port information.
+        """
+        if not self.trusted_publisher_url:
+            return False
+
+        publisher_uri = api.uri_reference(self.trusted_publisher_url).normalize()
+        user_uri = api.uri_reference(url).normalize()
+        is_subpath = publisher_uri.path == user_uri.path or user_uri.path.startswith(
+            publisher_uri.path + "/"
+        )
```
Comment (woodruffw) [inline warehouse/packaging/models.py:684]: NB: Probably need to check what happens when either URL doesn't have a `path` component; I suspect `path` will be `None` in that case 🙂

File: warehouse/packaging/models.py (hunk 4 above)
Comment (facutuesca) [reply warehouse/packaging/models.py]: fixed

File: warehouse/packaging/models.py (hunk 5 above)
Comment (facutuesca) [reply warehouse/packaging/models.py]: Good catch! Added a check, and a test

File: tests/unit/packaging/test_models.py [hunk 6]
```diff
@@ -555,6 +555,70 @@ def test_urls(self, db_session, home_page, download_url, project_urls, expected)
...
+            ),
+            (  # Host component matches, but contains user and port info
+                "https://user@github.com:443/owner/project",
+                "https://github.com/owner/project",
+                False,
+            ),
+            (  # URL path component is empty
+                "https://github.com",
+                "https://github.com/owner/project",
+                False,
```
Comment (woodruffw) [inline tests/unit/packaging/test_models.py:614]: A testcase for the inverse (TP has path, expected has no path) would also be good! And same for both having no path.

Comment (facutuesca) [reply tests/unit/packaging/test_models.py:614]: Done!
